Right click menus PLUS openTabs by container and discarded tab cleanup

kennethtran93 opened the pull request:
- Closes #717.
- Closes #521 in regards to adding expression via right-click-selection-text.
- Closes #415 in regards to right-click shortcuts.
- Closes #386 (Notifications on manual clean of cookies/localstorage from popup or right-click).
- Opening settings will now open next to the current tab.
- Reworked CleanupService tests.  The previous decreasing coverage was getting to me.

Updated 2020 June 05
- Closes #720 (Tab Discard), based on initial draft PR #721.
- Closes #716 and Closes #262 (Open Tabs per container).

kennethtran93 committed:
Initial Right-Click support. Still WIP.

Signed-off-by: Kenneth Tran <kennethtran93@users.noreply.github.com>

kennethtran93 committed:
WIP: Context Menu design

Signed-off-by: Kenneth Tran <kennethtran93@users.noreply.github.com>

kennethtran93 committed:
WIP:  Right-Click Context Menus. Per #717

- Redesigned Menu structure.
- Right-Click selected text to import as expression - closes #521
- Moved cleanDomainData functions from popup to Libs for reuse in menu.
- Enhanced cleanDomainData functions with notification and removal count.  Closes #386

Signed-off-by: Kenneth Tran <kennethtran93@users.noreply.github.com>

kennethtran93 committed:
With Release Notes

Signed-off-by: Kenneth Tran <kennethtran93@users.noreply.github.com>

kennethtran93 committed:
Working Right-Click ContextMenus.

Signed-off-by: Kenneth Tran <kennethtran93@users.noreply.github.com>

kennethtran93 committed:
Add Localization to Right-Click ContextMenus.

Signed-off-by: Kenneth Tran <kennethtran93@users.noreply.github.com>

kennethtran93 committed:
more localization notification adjustments.

Signed-off-by: Kenneth Tran <kennethtran93@users.noreply.github.com>

kennethtran93 committed:
fix wrong variable

Signed-off-by: Kenneth Tran <kennethtran93@users.noreply.github.com>

kennethtran93 committed:
Add Tests for new/revised functions.

Signed-off-by: Kenneth Tran <kennethtran93@users.noreply.github.com>

kennethtran93 commented:
Testers needed - right-click menu tests.

[Cookie-AutoDelete_Dev_20200601_181157_Chrome.zip](https://github.com/Cookie-AutoDelete/Cookie-AutoDelete/files/4712997/Cookie-AutoDelete_Dev_20200601_181157_Chrome.zip)
[Cookie-AutoDelete_Dev_20200601_181157_Firefox.zip](https://github.com/Cookie-AutoDelete/Cookie-AutoDelete/files/4712998/Cookie-AutoDelete_Dev_20200601_181157_Firefox.zip)

kennethtran93 commented:
@ke-d Can you give this a test run and see if it all works for you?  If you have any suggestions for improvement I'd like to hear it.

kennethtran93 committed:
Major rework of functions and tests for CleanupService.

- Attempted to implement a barebone firefox webextension api jest mock.o

Signed-off-by: Kenneth Tran <kennethtran93@users.noreply.github.com>

kennethtran93 committed:
whoops left console.info as is in setup.js

Signed-off-by: Kenneth Tran <kennethtran93@users.noreply.github.com>

kennethtran93 commented:
Realistically this build shouldn't have any drastic changes aside from the unit tests...

[Cookie-AutoDelete_Dev_20200605_032814_Chrome.zip](https://github.com/Cookie-AutoDelete/Cookie-AutoDelete/files/4733741/Cookie-AutoDelete_Dev_20200605_032814_Chrome.zip)

[Cookie-AutoDelete_Dev_20200605_032814_Firefox.zip](https://github.com/Cookie-AutoDelete/Cookie-AutoDelete/files/4733748/Cookie-AutoDelete_Dev_20200605_032814_Firefox.zip)

Feedback/Suggestions from testing is appreciated...before this gets mass-released.

kennethtran93 committed:
Add support to clean discarded tabs.

Open Tabs now take into consideration containers.

Refactored some code.

Signed-off-by: Kenneth Tran <kennethtran93@users.noreply.github.com>

kennethtran93 committed:
update typescript to 3.9.5 latest.

Signed-off-by: Kenneth Tran <kennethtran93@users.noreply.github.com>

kennethtran93 commented:
As it was somewhat easier to continue enhancements after refactoring the CleanupService tests, the completed enhancements has been included in this PR.

- Option to clean tabs that have tab.discarded set true.
- Bugfix: Open Tabs are now grouped by containers.

[Cookie-AutoDelete_Dev_20200605_194111_Chrome.zip](https://github.com/Cookie-AutoDelete/Cookie-AutoDelete/files/4738228/Cookie-AutoDelete_Dev_20200605_194111_Chrome.zip)
[Cookie-AutoDelete_Dev_20200605_194111_Firefox.zip](https://github.com/Cookie-AutoDelete/Cookie-AutoDelete/files/4738229/Cookie-AutoDelete_Dev_20200605_194111_Firefox.zip)

kennethtran93 committed:
Added release notes.

kennethtran93 committed:
Merge branch '3.X.X-Branch' into right-click-menus

ke-d commented on __tests__/services/CleanupService.spec.ts:
```diff
@@ -104,116 +150,644 @@ const mockCookie: CookiePropertiesCleanup = {
+        expect(global.browser.cookies.getAll).toHaveBeenCalledTimes(3);
+        expect(global.browser.cookies.remove).toHaveBeenCalledTimes(2);
+        expect(global.browser.browsingData.removeLocalStorage).not.toHaveBeenCalled();
+
```
To be honest, if some or all of the `toHaveBeenCalled();` lines can be removed without decreasing coverage, then I rather not have it to not clutter the tests.

The only time it makes sense is that if you are testing something important like `        expect(global.browser.cookies.getAll).toHaveBeenCalledTimes(3);` where the input would affect this in the CleanupService.

kennethtran93 replied:
I'll try and minimize the amount of tests, but coverage requires all functions to be covered, irregardless of where it affects the extension...

kennethtran93 replied:
As the previous test have already covered it, I'll remove redundancy ones.

kennethtran93 replied:
Should be resolved through 2b5e3b2

ke-d commented on __tests__/services/CleanupService.spec.ts:
```diff
@@ -104,116 +150,644 @@ const mockCookie: CookiePropertiesCleanup = {
+        // isFirstPartyIsolate plus one for each storeId: 'default', 'firefox-default'
+        expect(global.browser.cookies.getAll).toHaveBeenCalledTimes(3);
+        // this error always call cadLog even if debugMode is disabled.
+        expect(spyLib.cadLog).toHaveBeenCalledTimes(1);
```
I'm not sure if it's necessary to spy the cadlog in here and other places.

kennethtran93 replied:
With coverage, at least one test is needed to cover the functions called, including cadLog, depending on how the path to it is given.

ke-d commented on __tests__/setup.js:
```diff
@@ -0,0 +1,150 @@
+ * SOFTWARE.
+ */
+'use strict';
+
```
Can you add a comment explaining the purpose of this file?

kennethtran93 replied:
Should be resolved through 2b5e3b2

ke-d commented on src/background.ts:
```diff
@@ -151,14 +166,330 @@ const onStartUp = async () => {
+) {
+  const debug = getSetting(store.getState(), 'debugMode') as boolean;
+  const contextualIdentities = getSetting(store.getState(), 'contextualIdentities') as boolean;
+  if (debug) {
```
Instead of if (debug) statements everywhere, might I suggest just calling 

```
    cadLog({
      msg: `background.onContextMenuClicked:  Data received`,
      x: {info, tab},
    });
```
And then inside the cadLog function, it would have the `if (debug)` statement. Doing this should reduce the amount of test cases as well since there's no more condition.

kennethtran93 replied:
So what you're saying is to pass another variable into cadLog, call it debug, and then in each of the calling functions pass in ```getSetting(store.getState(), 'debugMode') as boolean```?

kennethtran93 replied:
Heads up - when I'm done working on this, more files will be changed.

Normally I'd say if a function doesn't need to be called then use `if`, but if you want to minimize test cases then we'll do it that way then.

ke-d replied:
> So what you're saying is to pass another variable into cadLog, call it debug, and then in each of the calling functions pass in getSetting(store.getState(), 'debugMode') as boolean?

Yeah that would work.

Eventually though we may want to do some dependency injection (i.e. make cadLog into a singleton class function that has the state of the extension) so we don't have to pass in  `getSetting(store.getState(), 'debugMode')`. But your solution will work for the short term.

kennethtran93 replied:
I'd haven't thought about moving cadLog from Libs to a separate class...I had thought to keep the multiple use functions into Libs.

Mind you this was only an initial debug log - I had intend for it to either show up on a new settings page for easier debugging/exporting.

Also in the future this debug should have levels as there is so much information being sent through.

kennethtran93 replied:
The only other thing is that there are some cadLog calls that still outputs regardless of debug mode, so if we're moving it into a separate class that needs to be taken into consideration.

kennethtran93 replied:
Should be resolved through 2b5e3b2, at least through adding in the parameter for now.

ke-d commented on src/background.ts:
```diff
@@ -151,14 +166,330 @@ const onStartUp = async () => {
+  }
 };
 
+async function onContextMenuClicked(
```
There's a lot of code here, can we move this function to another file?

kennethtran93 replied:
I'll move it into ContextMenuEvents.

kennethtran93 replied:
Should be resolved through 2b5e3b2

ke-d commented on src/redux/Actions.ts:
```diff
@@ -271,12 +271,12 @@ export const cookieCleanup: ActionCreator<
     const notifyMessage = browser.i18n.getMessage('notificationContent', [
       recentlyCleaned.toString(),
-      Array.from(setOfDeletedDomainCookies).join(', '),
+      setOfDeletedDomainCookies.join(', '),
```
Why is this needed?

kennethtran93 replied:
This was because when I was reading the output the cadLog gave, it only returned `{}` when passed in as a Set. Well, `returnSetOfOpenTabDomain` now returns an object with the container id as the key, then the open tab domains.    As such I modified it in `returnSetOfOpenTabDomains` to return an array instead of a Set.  

It doesn't necessarily need to be changed, but if an array is already being returned, there's no point of converting an array into an array again.

kennethtran93 replied:
Marking as resolved per above comment, unless changes in `returnSetOfOpenTabDomains` should be renamed now that it isn't returning a set...

kennethtran93 replied:
In 342cc64 I've renamed `returnSetOfOpenTabDomains` to `returnContainersOfOpenTabDomains`  
The function now returns an object, with the container id as key and an array of open tab domains.

ke-d reviewed (changes requested):
Thanks for the PR. Would recommend 1 PR per 1 issue resolved next time. :)

kennethtran93 committed:
Changes requested per ke-d.

- Moved contextMenu onClicked into its class from background.
- Adjusted some tests while keeping coverage same.
- Added:  Failed notification should be shown if right-click domain wasn't found

Signed-off-by: Kenneth Tran <kennethtran93@users.noreply.github.com>

kennethtran93 committed:
rename returnSetOfOpenTabDomains as it is not return Set anymore.

Signed-off-by: Kenneth Tran <kennethtran93@users.noreply.github.com>

ke-d commented on src/services/Libs.ts:
```diff
@@ -238,6 +257,13 @@ export const localFileToRegex = (hostname: string) => {
+/**
+ * Parse cookieStoreId for use in addExpressionUI...
+ */
+export const parseCookieStoreId = (contextualIdentities: boolean, cookieStoreId: string | undefined) => {
```
Minor nit: I think this can be `cookieStoreId: string?`

kennethtran93 replied:
While I'd like to, that is actually an optional parameter per https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/tabs/Tab, not to mention Chrome doesn't have one.  The check within the function checks if it is defined otherwise assigns 'default' when returning for CAD lists.

Also when I tried removing it just now TypeScript is complaining, even though we'll always have the appropriate firefox permission to always have cookieStoreId as a parameter.

ke-d commented on src/services/CleanupService.ts:
```diff
@@ -222,79 +205,154 @@ export const cleanCookies = async (
+        storeId: cookie.storeId,
+        url: prepareCookieDomain(cookie),
+      }, firstPartyIsolate) as {
+        // Fix type error with undefineds with cookies.remove
```
Interesting, I wonder why is this?

kennethtran93 replied:
What are you referring to here?

kennethtran93 replied:
I'll keep this PR/branch until we clarify what is needed here.

ke-d replied:
Why does `browser.cookies.remove` return undefined here? Is the return type not defined in the types library?

kennethtran93 replied:
I'm not sure - I had extracted the code that used to be in the popup, so whatever you had done at that time.

Upon further research, the `url` isn't defined in cookies as the url that is required in `cookies.remove` consists of the combination of domain + path that we take from cookies.  Because url and name is required, it doesn't take in `undefined` within the type, unlike our other custom types that were extended from `browser.cookies.Cookie` with optional parameters

kennethtran93 replied:
also because `returnOptionalCookieAPIAttributes` has an input type of `Partial<CookiePropertiesCleanup>` which I believe applies `| undefined` to all parameters.

ke-d commented on src/services/CleanupService.ts:
```diff
@@ -222,79 +205,154 @@ export const cleanCookies = async (
+        url: string;
+        name: string;
+      });
+      if (r !== null) cookieDeletedCount += 1;
```
`if (!r)` ?

kennethtran93 replied:
yeah I can change that.  I was just following MDN where it says it will return null if it wasn't removed.

kennethtran93 replied:
so I guess it should be `if (r)` instead.

kennethtran93 replied:
resolved via 6a1a285

ke-d reviewed (dismissed):
Some minor stuff and I don't see any other major problems so approved.

kennethtran93 committed:
one minor change for null

Signed-off-by: Kenneth Tran <kennethtran93@users.noreply.github.com>

kennethtran93 committed:
Add .idea directory to .gitignore

Signed-off-by: Kenneth Tran <kennethtran93@users.noreply.github.com>

wrkq commented:
Hi. Sorry for late comment.
The cleanup of cookies per container seems to work exactly as it should now, after testing the basic couple of scenarios. It does not seem to accidentally mess up cookies in whitelisted containers or anything like that either.
Thank you a ton for this. It's really nice to not have to manually clean up when that one dumb social media link or something opens in current container.
